### PR TITLE
Refactor Dockerfile

### PR DESCRIPTION
**Pull request by hughcapet**
Split the main Dockerfile into multiple .sh scripts for the sake of readability and an easier build failures handling

### CODE REVIEW

File: postgres-appliance/build_scripts/prepare.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+export DEBIAN_FRONTEND=noninteractive
+
+echo 'APT::Install-Recommends "0";\nAPT::Install-Suggests "0";' > /etc/apt/apt.conf.d/01norecommend
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/prepare.sh:5]:
```suggestion
echo -e 'APT::Install-Recommends "0";\nAPT::Install-Suggests "0";' > /etc/apt/apt.conf.d/01norecommend
```

File: postgres-appliance/build_scripts/walg.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
+
+## -------------------------
+## Install wal-g
+## -------------------------
+
+export DEBIAN_FRONTEND=noninteractive
+
+echo 'APT::Install-Recommends "0";\nAPT::Install-Suggests "0";' > /etc/apt/apt.conf.d/01norecommend
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/walg.sh:9]:
```suggestion
echo -e 'APT::Install-Recommends "0";\nAPT::Install-Suggests "0";' > /etc/apt/apt.conf.d/01norecommend
```

File: postgres-appliance/build_scripts/walg.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+apt-get update
+apt-get install -y curl ca-certificates
+
+if [ "$(dpkg --print-architecture)" != "amd64" ]; then
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/walg.sh:14]:
```suggestion
ARCH="$(dpkg --print-architecture)"
if [ "$ARCH" != "amd64" ]; then
```

and so on

File: postgres-appliance/build_scripts/walg.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+if grep -q DISTRIB_RELEASE=18.04 /etc/lsb-release; then export CGO_LDFLAGS=-no-pie; fi
+    export USE_LIBSODIUM=1
+    export USE_LZO=1
+    export MAKEFLAGS="-j $(grep -c ^processor /proc/cpuinfo)"
+    make pg_build
+fi
+
+# We want to remove all libgdal20 debs except one that is for current architecture.
+echo "shopt -s extglob\nrm /builddeps/!(*_$(dpkg --print-architecture).deb)" | bash -s
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/walg.sh:39]:
```suggestion
echo -e "shopt -s extglob\nrm /builddeps/!(*_$ARCH.deb)" | bash -s
```

**Review by CyberDem0n — COMMENTED** [review]
`echo` behavior differs between `sh` and `bash`

File: postgres-appliance/build_scripts/base.sh [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+            if [ "$v" != "/usr/lib/postgresql/$PGVERSION" ] && [ -f "$v/bin/$u" ]; then
+                rm "$v/bin/$u"
+                ln -s "../../$PGVERSION/bin/$u" "$v/bin/$u"
+            fi;
+        done;
+    done
+
+    set +x
+
+    for v1 in $(find /usr/share/postgresql -mindepth 1 -maxdepth 1 | sort -Vr); do
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/base.sh:265]:
```suggestion
    for v1 in $(find /usr/share/postgresql -type d -mindepth 1 -maxdepth 1 | sort -Vr); do
```

File: postgres-appliance/build_scripts/base.sh [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        cd "$v1/extension"
+        while IFS= read -r -d '' orig
+        do
+            for f in "${orig%.sql}"--*.sql; do
+                if [ ! -L "$f" ] && diff "$orig" "$f" > /dev/null; then
+                    echo "creating symlink $f -> $orig"
+                    rm "$f" && ln -s "$orig" "$f"
+                fi
+            done
+        done <  <(find . -maxdepth 1 -name '*.sql' -not -name '*--*')
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/base.sh:276]:
```suggestion
        done <  <(find . -type f -maxdepth 1 -name '*.sql' -not -name '*--*')
```

File: postgres-appliance/build_scripts/base.sh [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+                        echo "creating symlink $f -> $orig"
+                        rm "$f" && ln -s "$orig" "$f"
+                    fi
+                done
+            fi
+        done
+
+        # relink files with the same name and content across different major versions
+        started=0
+        for v2 in $(find /usr/share/postgresql -mindepth 1 -maxdepth 1 | sort -Vr); do
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/base.sh:292]:
```suggestion
        for v2 in $(find /usr/share/postgresql -type d -mindepth 1 -maxdepth 1 | sort -Vr); do
```

File: postgres-appliance/build_scripts/compress_build.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ln -snf busybox /bin/sh
+
+files="/bin/sh"
+arch=$(uname -m)
+darch=$(uname -m | sed 's/_/-/')
+libs="$(ldd $files | awk '{print $3;}' | grep '^/' \
+    | sort -u) /lib/ld-linux-$darch.so.* /lib/$arch-linux-gnu/ld-linux-$darch.so.* /lib/$arch-linux-gnu/libnsl.so.* \
+    /lib/$arch-linux-gnu/libnss_compat.so.*"
+
+(echo /var/run /var/spool "$files" "$libs" | tr ' ' '\n' && realpath "$files" "$libs") | sort -u | sed 's/^\///' > /exclude
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/compress_build.sh:19]: Quoted "$libs" disables name globbing, but we rely on it.

```suggestion
libs=("$(ldd $files | awk '{print $3;}' | grep '^/' \
    | sort -u)" /lib/ld-linux-$darch.so.* /lib/$arch-linux-gnu/ld-linux-$darch.so.* /lib/$arch-linux-gnu/libnsl.so.* \
    /lib/$arch-linux-gnu/libnss_compat.so.*)

(echo /var/run /var/spool "$files" "${libs[@]}" | tr ' ' '\n' && realpath "$files" "${libs[@]}") | sort -u | sed 's/^\///' > /exclude
```

File: postgres-appliance/build_scripts/compress_build.sh [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+find /etc/alternatives -xtype l -delete
+save_dirs="usr lib var bin sbin etc/ssl etc/init.d etc/alternatives etc/apt"
+XZ_OPT=-e9v tar -X /exclude -cpJf a.tar.xz "$save_dirs"
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/compress_build.sh:23]:
```suggestion
save_dirs=(usr lib var bin sbin etc/ssl etc/init.d etc/alternatives etc/apt)
XZ_OPT=-e9v tar -X /exclude -cpJf a.tar.xz "${save_dirs[@]}"
```

File: postgres-appliance/build_scripts/compress_build.sh [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+rm -fr /usr/local/lib/python*
+
+/bin/busybox sh -c "(find $save_dirs -not -type d && cat /exclude /exclude && echo exclude) | sort | uniq -u | xargs /bin/busybox rm"
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/compress_build.sh:27]:
```suggestion
/bin/busybox sh -c "(find ${save_dirs[@]} -not -type d && cat /exclude /exclude && echo exclude) | sort | uniq -u | xargs /bin/busybox rm"
```

File: postgres-appliance/build_scripts/compress_build.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+/bin/busybox --install -s
+/bin/busybox sh -c "find $save_dirs -type d -depth -exec rmdir -p {}; 2> /dev/null"
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/compress_build.sh:29]:
```suggestion
/bin/busybox sh -c "find ${save_dirs[@]} -type d -depth -exec rmdir -p {}; 2> /dev/null"
```

File: postgres-appliance/build_scripts/patroni_wale.sh [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+# install most of the patroni dependencies from ubuntu packages
+apt-cache depends patroni \
+        | sed -n -e 's/.* Depends:(python3-.\+\)$/\1/p' \
+        | grep -Ev '^python3-(sphinx|etcd|consul|kazoo|kubernetes)' \
+        | xargs apt-get install -y "${BUILD_PACKAGES[@]}" python3-pystache python3-requests
+
+pip3 install setuptools
+
+if [ "$DEMO" != "true" ]; then
+    export EXTRAS=",etcd,consul,zookeeper,aws"
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/patroni_wale.sh:24]: here we don't need to export
```suggestion
    EXTRAS=",etcd,consul,zookeeper,aws"
```

File: postgres-appliance/build_scripts/post_build.sh [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+chown -R postgres:postgres "$PGHOME" "$RW_DIR"
+
+rm -fr /var/spool/cron /var/tmp
+mkdir -p /var/spool
+ln -s "$RW_DIR/cron" /var/spool/cron
+ln -s "$RW_DIR/tmp" /var/tmp
+
+for d in /etc/service/*; do
+    chmod 755 "$d/*"
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/post_build.sh:13]: Name globbing is important
```suggestion
    chmod 755 "$d"/*
```

File: postgres-appliance/build_scripts/walg.sh [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+if [ "$DEMO" = "true" ]; then
+    rm -f /builddeps/*.deb
+    # Create an empty dummy deb file to prevent the `COPY --from=dependencies-builder /builddeps/*.deb /builddeps/` step from failing
+    touch /builddeps/dummy.deb
+elif [ "$ARCH" != "amd64" ]; then
+    cp /wal-g/main/pg/wal-g /builddeps/wal-g/
+else
+    # In order to speed up amd64 build we just download the binary from GH
+    DISTRIB_RELEASE=$(sed -n 's/DISTRIB_RELEASE=//p' /etc/lsb-release)
+    curl -sL https://github.com/wal-g/wal-g/releases/download/"$WALG_VERSION"/wal-g-pg-ubuntu-"$DISTRIB_RELEASE"-amd64.tar.gz \
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/walg.sh:55]:
```suggestion
    curl -sL "https://github.com/wal-g/wal-g/releases/download/$WALG_VERSION/wal-g-pg-ubuntu-$DISTRIB_RELEASE-amd64.tar.gz" \
```

File: postgres-appliance/build_scripts/walg.sh [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+                | tar -C /builddeps/wal-g -xz
+    mv /builddeps/wal-g/wal-g-pg-ubuntu-"$DISTRIB_RELEASE"-amd64 /builddeps/wal-g/wal-g
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/walg.sh:57]:
```suggestion
    mv "/builddeps/wal-g/wal-g-pg-ubuntu-$DISTRIB_RELEASE-amd64" /builddeps/wal-g/wal-g
```

File: postgres-appliance/build_scripts/walg.sh [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    touch /builddeps/dummy.deb
+elif [ "$ARCH" != "amd64" ]; then
+    cp /wal-g/main/pg/wal-g /builddeps/wal-g/
+else
+    # In order to speed up amd64 build we just download the binary from GH
+    DISTRIB_RELEASE=$(sed -n 's/DISTRIB_RELEASE=//p' /etc/lsb-release)
+    curl -sL "https://github.com/wal-g/wal-g/releases/download/$WALG_VERSION/wal-g-pg-ubuntu-$DISTRIB_RELEASE-amd64.tar.gz" \
+                | tar -C /builddeps/wal-g -xz
+    mv "/builddeps/wal-g/wal-g-pg-ubuntu-$DISTRIB_RELEASE-amd64" /builddeps/wal-g/wal-g
+fi
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/walg.sh:58]: This one could probably be a different build layer.
It is being abused to remove one of: (libgdal20_2.2.3+dfsg-2_amd64.deb, libgdal20_2.2.3+dfsg-2_arm64.deb).

On a side note, `ARG DEMO` is lost here: the https://github.com/zalando/spilo/pull/746/files#diff-8918d403d0c48088cebb6ed78df4fecd856e211f7522e06ee423383b17bd8453L39

File: postgres-appliance/build_scripts/base.sh [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+export DEBIAN_FRONTEND=noninteractive
+MAKEFLAGS="-j $(grep -c ^processor /proc/cpuinfo)"
+export MAKEFLAGS
+
+set -ex
+sed -i 's/^#\s*\(deb.*universe\)$/\1/g' /etc/apt/sources.list
+
+apt-get update
+
+cd /builddeps
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/base.sh:16]: Could be translated to the WORKDIR in the Dockerfile

File: postgres-appliance/build_scripts/base.sh [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+            pg_test_timing \
+            pgbench \
+            reindexdb \
+            vacuumlo *.py; do
+        for v in /usr/lib/postgresql/*; do
+            if [ "$v" != "/usr/lib/postgresql/$PGVERSION" ] && [ -f "$v/bin/$u" ]; then
+                rm "$v/bin/$u"
+                ln -s "../../$PGVERSION/bin/$u" "$v/bin/$u"
+            fi;
+        done;
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/base.sh:263]:
```suggestion
            fi
        done
```

File: postgres-appliance/build_scripts/base.sh [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+        /etc/pgbouncer/* \
+        /usr/lib/postgresql/*/bin/createdb \
+        /usr/lib/postgresql/*/bin/createlang \
+        /usr/lib/postgresql/*/bin/createuser \
+        /usr/lib/postgresql/*/bin/dropdb \
+        /usr/lib/postgresql/*/bin/droplang \
+        /usr/lib/postgresql/*/bin/dropuser \
+        /usr/lib/postgresql/*/bin/pg_standby \
+        /usr/lib/postgresql/*/bin/pltcl_*
+find /var/log -type f -exec truncate --size 0 {} \;
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/base.sh:343]:
```suggestion
find /var/log -type f -exec truncate --size 0 {}

File: postgres-appliance/build_scripts/prepare.sh [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+apt-get purge -y libcap2-bin
+apt-get autoremove -y
+apt-get clean
+rm -rf /var/lib/apt/lists/* \
+            /var/cache/debconf/* \
+            /usr/share/doc \
+            /usr/share/man \
+            /usr/share/locale/?? \
+            /usr/share/locale/??_??
+find /var/log -type f -exec truncate --size 0 {} \;
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/prepare.sh:64]:
```suggestion
find /var/log -type f -exec truncate --size 0 {}
```

File: postgres-appliance/build_scripts/patroni_wale.sh [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+apt-get clean
+rm -rf /var/lib/apt/lists/* \
+        /var/cache/debconf/* \
+        /root/.cache \
+        /usr/share/doc \
+        /usr/share/man \
+        /usr/share/locale/?? \
+        /usr/share/locale/??_?? \
+        /usr/share/info
+find /var/log -type f -exec truncate --size 0 {} \;
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/patroni_wale.sh:84]:
```suggestion
find /var/log -type f -exec truncate --size 0 {}
```

File: postgres-appliance/build_scripts/base.sh [hunk 22]
```diff
@@ -0,0 +1,343 @@
...
+    done
+fi
+
+    # build and install missing packages
+for pkg in pgextwlist "${MISSING[@]}" "${MISSING_EXTRAS[@]}"; do
+    apt-get source "postgresql-14-${pkg}"
+    # use subshell to avoid having to cd back (SC2103)
+    (
+        cd "$(ls -d -- *"${pkg%?}"*-*/)"
+        if [ -f ../"$pkg.patch" ]; then patch -p1 < ../"$pkg".patch; fi
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/base.sh:205]:
```suggestion
        if [ -f "../$pkg.patch" ]; then patch -p1 < "../$pkg".patch; fi
```

File: postgres-appliance/build_scripts/compress_build.sh [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+files="/bin/sh"
+arch=$(uname -m)
+darch=$(uname -m | sed 's/_/-/')
+
+IFS=" " read -r -a libs <<< "$(ldd $files | awk '{print $3;}' | grep '^/' | sort -u)"
+libs+=(/lib/ld-linux-"$darch".so.* \
+    /lib/"$arch"-linux-gnu/ld-linux-"$darch".so.* \
+    /lib/"$arch"-linux-gnu/libnsl.so.* \
+    /lib/"$arch"-linux-gnu/libnss_compat.so.*)
+echo "${libs[@]}"
```
Comment (CyberDem0n) [inline postgres-appliance/build_scripts/compress_build.sh:21]: Debug?

Comment (hughcapet) [reply postgres-appliance/build_scripts/compress_build.sh:21]: 🤦‍♀️ again

File: postgres-appliance/build_scripts/base.sh (hunk 19 above)
Comment (hughcapet) [reply postgres-appliance/build_scripts/base.sh:343]: man find
> **-exec command ;** - <...> All following arguments to find are taken to be arguments to the command until an argument consisting of ';' is  encountered. <...> might need to be escaped (with a '\\') or quoted to protect them from expansion by the shell

Comment (CyberDem0n) [reply postgres-appliance/build_scripts/base.sh:343]: sorry, totally forgot about find syntax :man_facepalming:

File: postgres-appliance/build_scripts/walg.sh (hunk 16 above)
Comment (hughcapet) [reply postgres-appliance/build_scripts/walg.sh]: I better just rename the file :) Hope you have no objection

Comment (CyberDem0n) [reply postgres-appliance/build_scripts/walg.sh]: I am fine with it

**Comment by CyberDem0n** [comment]
:+1:

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Comment by hughcapet** [comment]
👍

**Pull request merged**